File: 백준/Silver/3986. 좋은 단어/좋은 단어.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().split("\n");

//교차만 안하면됨 그렇다면 stack 에 다 담아놓고 이전거와 바로 비교하면됨

const N = Number(input[0])

let count = 0;

// 100인풋은 100이고 범위는 100,000 100 x 100,000 루프니까 1초 딱 됨
// O( N x M) 시간 복잡도를 가진다고 보면된다.
for (let i = 1; i <= N; i++) {
  const stack = [];
  let str = input[i].split("")

  for (let char of str) {
    if (stack.length > 0 && stack[stack.length - 1] === char) {
      stack.pop()
    } else {
      stack.push(char)
    }
  }

  if (stack.length === 0) count++;
}

console.log(count)

File: 백준/Silver/3986. 좋은 단어/좋은 단어.py
import sys
from collections import deque 

N = int(sys.stdin.readline())
count = 0

for _ in range(N):
  str = sys.stdin.readline().rstrip()
  stack = deque()

  for char in str:
    if stack and stack[-1] == char:
      stack.pop()
    else:
      stack.append(char)


  if not stack:
    count +=1

print(count)

